crates/pallet-domains/src/staking_epoch.rs

```diff
@@ -0,0 +1,884 @@
+            let nominator_share = Perbill::from_rational(nominator.shares, total_shares);
+            let nominator_staked_amount = nominator_share.mul_floor(total_stake);
+
+            let locked_amount = T::Currency::balance_frozen(&freeze_identifier, &nominator_id);
+            let amount_to_mint = nominator_staked_amount
+                .checked_sub(&locked_amount)
+                .unwrap_or(Zero::zero());
+
+            // remove the lock and mint any gains
+            T::Currency::thaw(&freeze_identifier, &nominator_id)
+                .map_err(|_| TransitionError::RemoveLock)?;
+
+            mint_funds::<T>(&nominator_id, amount_to_mint)?;
+
+            // update pool's remaining shares and stake
+            total_shares = total_shares
+                .checked_sub(&nominator.shares)
+                .ok_or(TransitionError::ShareUnderflow)?;
+            total_stake = total_stake
+                .checked_sub(&nominator_staked_amount)
+                .ok_or(TransitionError::BalanceUnderflow)?;
+
+            // remove any pending deposits for this nominator and then remove the lock
+            PendingDeposits::<T>::remove(operator_id, nominator_id);
+
+            Ok(())
+        })?;
+
+        // TODO: transfer any remaining amount to treasury
+
+        // remove all of the pending deposits since we initiated withdrawal for all nominators.
+        let _ = PendingWithdrawals::<T>::clear_prefix(operator_id, u32::MAX, None);
+
+        // remove lock on any remaining deposits, all these are new nominators recorded after start
+        // of new epoch and before operator de-registered
+        for (nominator_id, _) in PendingDeposits::<T>::drain_prefix(operator_id) {
+            T::Currency::thaw(&freeze_identifier, &nominator_id)
+                .map_err(|_| TransitionError::RemoveLock)?;
+        }
+
+        // remove OperatorOwner Details
+        OperatorIdOwner::<T>::remove(operator_id);
+
+        Ok(())
+    })
+    .map_err(Error::UnlockOperator)
+}
+
+fn unlock_nominator_withdrawals<T: Config>(
+    operator_id: OperatorId,
+    current_consensus_block: T::BlockNumber,
+) -> Result<(), Error> {
+    let freeze_identifier = T::FreezeIdentifier::staking_freeze_id(operator_id);
+    match PendingNominatorUnlocks::<T>::take(operator_id, current_consensus_block) {
+        None => Ok(()),
+        Some(withdrawals) => withdrawals.into_iter().try_for_each(|withdrawal| {
+            let frozen_balance =
+                T::Currency::balance_frozen(&freeze_identifier, &withdrawal.nominator_id);
+
+            let remaining_staked_balance = frozen_balance
+                .checked_sub(&withdrawal.balance)
+                .ok_or(TransitionError::BalanceUnderflow)?;
+
+            T::Currency::set_freeze(
+                &freeze_identifier,
+                &withdrawal.nominator_id,
+                remaining_staked_balance,
+            )
+            .map_err(|_| TransitionError::UpdateLock)?;
```

Comment:
If `remaining_staked_balance` is zero then we should not set freeze anymore.

Reply:
This is implicit assumption here